### PR TITLE
drivers: cc110x: cleanup, some bugs fixed, CCA support

### PR DESCRIPTION
**Pull request by kaspar030**
Here are some cc110x improvements I've been working on.
Most notable changes are a reworked interrupt logic, which eliminates a bug where nodes would get stuck in RX_BUSY after dropping a packet with the wrong sender address.

I've also implemented clear channel assessment using a very simple random-time based retry logic and proper error reporting (at least to netdev).

Tested on msba2. To make testing on msba2 easier, I highly recommend #3944, it saves a lot of address changing.

### CODE REVIEW

**Comment by kaspar030** [comment]
- rebased
- fixed some CCA bugs

**Comment by OlegHahm** [comment]
Sorry for the delayed review. Please ping me next Monday.

**Comment by zhuoshuguo** [comment]
@kaspar030 Ping, a kind reminder for rebase.

**Comment by kaspar030** [comment]
- rebased

**Comment by OlegHahm** [comment]
@kaspar030, can you remind me at Hack & ACK tomorrow to test this?

**Comment by kaspar030** [comment]
@OlegHahm Sure, though I won't be there ;(

**Comment by OlegHahm** [comment]
Understandable. ;-) I will have a beer on you and may ACK and merge this PR.

File: Makefile.dep [hunk 1]
```diff
@@ -429,3 +429,8 @@ endif
 ifneq (,$(filter ethos,$(USEMODULE)))
     USEMODULE += netdev2_eth
 endif
+
+ifneq (,$(filter cc110x,$(USEMODULE)))
+    USEMODULE += xtimer random
+    DISABLE_MODULE += gnrc_sixlowpan_iphc
```
Comment (OlegHahm) [inline Makefile.dep]: What was the rationale to disable IPHC?

File: drivers/cc110x/cc110x-rxtx.c [hunk 2]
```diff
@@ -117,46 +132,47 @@ static void _rx_read_data(cc110x_t *dev, void(*callback)(void*), void*arg)
         pkt_buf->lqi = status[I_LQI] & LQI_EST;
 
         /* MSB of LQI is the CRC_OK bit */
-        int crc_ok = (status[I_LQI] & CRC_OK) >> 7;
+        int crc_ok = status[I_LQI] & CRC_OK;
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: Can you keep the parens?

File: drivers/cc110x/cc110x-rxtx.c [hunk 3]
```diff
@@ -117,46 +132,47 @@ static void _rx_read_data(cc110x_t *dev, void(*callback)(void*), void*arg)
...
-                    LOG_DEBUG("cc110x: received packet from=%u to=%u payload "
-                            "len=%u\n",
+            LOG_DEBUG("cc110x: received packet from=%u to=%u payload "
+                    "len=%u\n",
                     (unsigned)pkt_buf->packet.phy_src,
                     (unsigned)pkt_buf->packet.address,
                     pkt_buf->packet.length-3);
+
+            /* count successful packet */
+            dev->cc110x_statistic.packets_in++;
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: It's about time for netstats. ;)

File: drivers/cc110x/cc110x-rxtx.c [hunk 4]
```diff
@@ -117,46 +132,47 @@ static void _rx_read_data(cc110x_t *dev, void(*callback)(void*), void*arg)
...
-    }
-
-    gpio_irq_disable(dev->params.gdo2);
-
-    do {
-        _rx_read_data(dev, callback, arg);
+        /* we're done. */
+        _rx_stop(dev);
+
+    } else {
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: style

File: drivers/cc110x/cc110x-rxtx.c [hunk 5]
```diff
@@ -117,46 +132,47 @@ static void _rx_read_data(cc110x_t *dev, void(*callback)(void*), void*arg)
...
+        left -= to_read;
+        if (left > 32) {
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: magic number?

File: drivers/cc110x/cc110x-rxtx.c [hunk 6]
```diff
@@ -117,46 +132,47 @@ static void _rx_read_data(cc110x_t *dev, void(*callback)(void*), void*arg)
...
+            /* assert gdo2 when RX FIFO over threshold, deassert when below */
+            cc110x_write_reg(dev, CC110X_IOCFG2, 0x0);
+        } else {
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: style

File: drivers/cc110x/cc110x-rxtx.c [hunk 7]
```diff
@@ -244,6 +285,33 @@ void cc110x_isr_handler(cc110x_t *dev, void(*callback)(void*), void*arg)
...
+
+    /* set GDO2 to CCA */
+    cc110x_write_reg(dev, CC110X_IOCFG2, 0x09);
+
+    /* wait until RSSI value is stable */
+    do {
+        cc110x_readburst_reg(dev, CC110X_PKTSTATUS, &regtmp, 1);
+    } while (! (regtmp & 0x50)); /* bit6 = CS, bit4 = CCA */
+
+    return (regtmp & 0x10); /* CCA */
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: This function also seems to include a couple of magic numbers.

File: drivers/cc110x/cc110x-rxtx.c [hunk 8]
```diff
@@ -278,11 +345,23 @@ int cc110x_send(cc110x_t *dev, cc110x_pkt_t *packet)
...
+            break;
+        }
+    }
 
-    cc110x_write_reg(dev, CC110X_IOCFG2, 0x02);
+    if(retries == CC110X_CCA_MAXTRIES) {
+        DEBUG("cc110x: cca failed.\n");
+        cc110x_switch_to_rx(dev);
+        return -EAGAIN;
+    } else {
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: style

File: drivers/cc110x/include/cc110x-internal.h [hunk 9]
```diff
@@ -61,6 +61,10 @@ extern "C" {
 #endif
 #define CC110X_MIN_CHANNR           (0)     /**< lowest possible channel number */
 #define CC110X_MAX_CHANNR           (0)     /**< highest possible channel number */
+#define CC110X_CCA_MAXTRIES         (10)    /**< Max TX retries if CCA not clear */
```
Comment (OlegHahm) [inline drivers/cc110x/include/cc110x-internal.h]: Shouldn't that be configurable via `NETOPT_CSMA_RETRIES`?

**Comment by OlegHahm** [comment]
Sorry cannot test - I only have one mini-USB cable with me.

**Comment by OlegHahm** [comment]
How can I test? Using gnrc_networking doesn't seem to work out of the box, because nodes will have the same IPv6 address.

**Comment by OlegHahm** [comment]
Needs a rebase.

File: drivers/cc110x/cc110x-rxtx.c [hunk 10]
```diff
@@ -278,11 +345,23 @@ int cc110x_send(cc110x_t *dev, cc110x_pkt_t *packet)
     gpio_irq_disable(dev->params.gdo2);
     dev->radio_state = RADIO_TX_BUSY;
 
-#ifdef MODULE_CC110X_HOOKS
-    cc110x_hook_tx();
-#endif
+    unsigned retries;
+    for (retries = 0; retries < CC110X_CCA_MAXTRIES; retries++) {
+        xtimer_usleep(CC110X_CCA_RETRY_DELAY * (genrand_uint32() >> 22));
```
Comment (OlegHahm) [inline drivers/cc110x/cc110x-rxtx.c]: `s/genrand/random/`

**Comment by OlegHahm** [comment]
I tried with https://github.com/RIOT-OS/RIOT/pull/3944, but still couldn't get it to fly.

File: Makefile.dep (hunk 1 above)
Comment (kaspar030) [inline Makefile.dep]: It didn't work, probably because of misformed link-local addresses. Have to verify if that is still the case.

**Comment by kYc0o** [comment]
@kaspar030 have you addressed Oleg's comments? You think this PR still a bug and it's necessary for the release?

**Comment by kYc0o** [comment]
@kaspar030 ping! :D

**Comment by PeterKietzmann** [comment]
@kaspar030 I would test this PR tonight if you rebase it

**Comment by kYc0o** [comment]
@kaspar030 can you squash this one too? Maybe @PeterKietzmann still wants to test it ;)

**Comment by kaspar030** [comment]
Last time I tried with @OlegHahm, this didn't work at all. So I'll have to take another look... -> postponed.

**Comment by kYc0o** [comment]
Ok, let's see later.

**Comment by miri64** [comment]
How about this release?

**Comment by OlegHahm** [comment]
@kaspar030, any chance for 2017.01?

**Comment by haukepetersen** [comment]
@kaspar030: could you do a quick rebase, I'd like to try this PR together with the remodeled SPI driver of the `msba2` - the current version of the `cc110x` driver seems broken and before I debug the driver it might be a good chance that this PR already solves something...

**Comment by OlegHahm** [comment]
The last time @kaspar030 and me tested this PR, it was broken.

**Comment by aabadie** [comment]
> The last time @kaspar030 and me tested this PR, it was broken.

Is it still the case. This is quite an old one, it needs rebase and some renaming (netdev2->netdev)

**Comment by miri64** [comment]
No significant progress for over a year. Closing as memo for now.

**Pull request closed without merging**